Rs485 APIs support

anthonytex opened the pull request:
After one month of inactivity and considering that the original author hasn't answered anymore i'm taking up his work: his original patch has been maintained.

ref: https://github.com/zephyrproject-rtos/zephyr/pull/35708

kumarnac committed:
uart: Add support for RS485 APIs

Adds support for configuring and enabling RS485 signalling
on supported UARTs.

Signed-off-by: Nachiketa Kumar <nachiketa.kumar@intel.com>

Signed-off-by: Antonio Tessarolo <anthonytexdev@gmail.com>

MaureenHelm commented:
@anthonytex Thanks for picking this up. Do you plan to add a driver implementation to this PR?

cc: @kumarnac

dcpleung commented:
The API and DTS changes look good. Though it would be better if there is an accompanied driver changes so we can see how it will be used.

mbolivar-nordic reviewed (changes requested):
Please squash the fixup commit properly so the patch series can be read in order. I agree a driver would be useful for evaluating the binding changes.

anthonytex commented:
@mbolivar-nordic Hello, can you please explain me what do you mean? Should I squash all in a single commit? Thanks

Regarding the driver: I could modify a driver to add rs485 half duplex support.

mbolivar-nordic commented:
> @mbolivar-nordic Hello, can you please explain me what do you mean? Should I squash all in a single commit? Thanks

In this case, yes, it should be a single commit.

In general, the upstream Zephyr development process is to structure your commits into a series of logical changes. By "logical change" I mean "each commit does one thing". This precludes doing things like "add X binding" in commit 1, and then "whoops, fix X binding" in commit 2: you should just make one commit that adds X correctly. But that doesn't mean you should only ever submit one commit. Splitting up commits into logical changes is a good thing, because it makes PRs easier to review by reading them one commit at a time, which many reviewers do.

(The contributing guide in the documentation tries to explain all this, but it does a terrible job in my opinion. In general, though, we do the same thing as Linux, where we send multiple revisions of a single patch series, except we force push into the same PR instead of sending another version to the mailing lists.)

> Regarding the driver: I could modify a driver to add rs485 half duplex support.

Yes, could you please? I'll outline my concerns here a bit more just to be specific:

- this is adding a **lot** of properties to the binding, not all of which seem to be generic to RS485 in general, so my worry is that many of them don't belong in this binding, and instead should be in some sort of child binding for a particular device
- the API changes don't help me decide whether that's true or not, because they seem limited to just "turn on RS485", which doesn't require any sort of turn around times, etc.
- Linux has a generic rs485 binding which is very different from this one: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml
- we try to share bindings with Linux when it makes sense (and please read https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings in detail if you haven't already, since you are submitting a binding upstream), so the divergence here is an issue, or at least a "smell"

I think a driver which actually used the new properties would help all of this this make more sense.

If you can provide a binding which 1. only adds things that are actually used by drivers, and 2. is compatible with the Linux binding, I will be very happy to approve it. If that's not possible, we can figure out next steps.

mbolivar-nordic commented:
> The [...] DTS changes look good.

@dcpleung please take note of my above comment for future reference when looking at DTS changes.

dcpleung commented:
> > The [...] DTS changes look good.
> 
> @dcpleung please take note of my above comment for future reference when looking at DTS changes.

For my own reference, do we want to track Linux's devicetree definitions? and to be as close to Linux as possible? I understand that it would probably make it easier to port device drivers, but just wish to make sure.

anthonytex commented:
>
> In this case, yes, it should be a single commit.
>
I agree with you, generally i would never ever split 2 commits like this, but i did it, this time, only to keep the original commit from the old PR. I will squash them so, no problem.

> * this is adding a **lot** of properties to the binding, not all of which seem to be generic to RS485 in general, so my worry is that many of them don't belong in this binding, and instead should be in a binding for a particular device with a specific compatible
> * the API changes don't help me decide whether that's true or not, because they seem limited to just "turn on RS485", which doesn't require any sort of turn around times, etc.

My first impression was like yours, because in my applications i never use the RS485 in full-duplex mode, but full-duplex mode is a possibility: so the control of receiver enable may be needed. 
Also i don't agree with the fact that property like *assertion/deassertion* don't belong to RS485: the control time of de/re assertion/deassertion it's critical to achieve line stability with high capacity long cables or star topology connections.
This is why also in linux there is *rs485-rts-delay*.
I've changed the original *turnaround* to *re-assertion/deassertion* to stay more close to the common names of transceiver signals. 

> * Linux has a generic rs485 binding which is very different from this one: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/serial/rs485.yaml
> * we try to share bindings with Linux when it makes sense (and please read https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings in detail if you haven't already, since you are submitting a binding upstream), so the divergence here is an issue, or at least a "smell"

I completely agree with the fact that staying close to linux bindings helps **a lot** of users (like me) for a lot of reasons. Linux has a generic binding here but it's limited to half-duplex mode and blinks an eye to software rs485 mode, so if you want to support **only** half-duplex (that is the most used configuration in the real world) i will delete some bindings, but, in the previous PR there was a  mention to the fact that linux's bindings were not complete.

Also and more important **i've also some concerns** i would like to discuss (remember the fact that i've picked up these APIs):
1. nanoseconds time scale is maybe more than what one need but **maybe** some hardware peripheral that have hardware rs485 mode might support this precision.
2. If i want to put an UART driver that doesn't have RS485 hardware mode (most), but i want to achieve it with software? Generally Linux drivers use UART control signals like RTS/CTS to control the transceiver: is it better to specify to use for example RTS for DE in the description? 
3. If one doesn't have RTS/CTS and want to use RS485, in linux one can specify (if the driver supports it obviously) `rts-gpios=<some>;` but here? (of course this is a lack of serial binding not the new one)

mbolivar-nordic commented:
> > > The [...] DTS changes look good.
> > 
> > 
> > @dcpleung please take note of my above comment for future reference when looking at DTS changes.
> 
> For my own reference, do we want to track Linux's devicetree definitions? and to be as close to Linux as possible? I understand that it would probably make it easier to port device drivers, but just wish to make sure.

Yes, we aim for compatibility with Linux whenever it makes sense. This is an explicitly documented design goal: https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems

For the details, please see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings.

mbolivar-nordic commented:
> I agree with you, generally i would never ever split 2 commits like this, but i did it, this time, only to keep the original commit from the old PR. I will squash them so, no problem.

Thanks. In situations like this, you can just add your sign-off line to the commit when squashing if you're making trivial changes. You can change the author line if you feel you are changing the commit substantially enough that you are the author. Either way, keep both authors' Signed-off-by: lines in the resulting squashed commit.

> I completely agree with the fact that staying close to linux bindings helps a lot of users (like me) for a lot of reasons. Linux has a generic binding here but it's limited to half-duplex mode and blinks an eye to software rs485 mode, so if you want to support only half-duplex (that is the most used configuration in the real world) i will delete some bindings, but, in the previous PR there was a mention to the fact that linux's bindings were not complete.

OK, that sounds like a good justification for a divergence from Linux -- like I wrote earlier, we aim for compatibility "whenever it makes sense". If you can justify the differences, I personally don't have any objections to them, but cc'ing @galak as well.

anthonytex committed:
uart: modify uart_imx driver to support RS485 API

NXP imx uart driver has been modified to include the new API for Rs485.
This driver support "software" half-duplex RS485 by using CTS pin
for DE (in DCE mode).

Signed-off-by: Antonio Tessarolo <anthonytexdev@gmail.com>

anthonytex commented:
So i've pushed a simple (draft) implementation with some assumptions and i've also squashed the 2 commits:
1. if one would use de timeout must have a tick time that satisfy requested precision (obviously with software RS485 one can expect milliseconds precision **not** nanosecond see linux API)
2. It's mandatory to call `int uart_irq_tx_complete(const struct device *dev)` with this driver after sending the the last byte to deactivate the complete IRQ.
3. ideally one user would call uart_irq_tx_complete_enable (but there is no such API) after sending the last byte to the FIFO.

GeorgeCGV commented:
@anthonytex, wouldn't it be better to have

```
rs485-de-active-low
rs485-re-active-high
```

as a single enum choice?

For example:
```
polarity:
  type: string
  required: true
  description: Configures driver enable signal active polarity.
      enum:
         - "active-low"
         - "active-high"
```
that avoids the possibility to define both polarities at the same time.

I am not sure regarding the de (de)assertion time being in nanoseconds in the bindings.
The STM32 documentation defines its own unit:

> The assertion and de-assertion values are expressed in term of simple time unit (stu) which
depends on the data oversampling mode. 

It might be the case with others that the time unit is not in nanoseconds.

anthonytex commented:
Hello @GeorgeCGV, the two bools refer to two different signals:
* one is driver enable
* the other is receiver enable

that's why they are separated.

> I am not sure regarding the (de)assertion time being in nanoseconds in the bindings. The STM32 documentation defines its unit:
> 
> > The assertion and de-assertion values are expressed in terms of simple time unit (STU) which
> > depends on the data oversampling mode.
> 
> It might be the case with others that the time unit is not in nanoseconds. Therefore, using `ns` in the bindings might cause misunderstandings or potential errors.


As you can see from my previous comment i've some concerns about the time scale:
> 1. nanoseconds time scale is maybe more than what one need but **maybe** some hardware peripheral that have hardware rs485 mode might support this precision.

But I believe that it's correct to have a time unit like ms, us or yes ns, so sub-seconds scale. 
In linux there is ms time unit but i think is not enough accurate. 
If a peripheral has a different unit type it must convert to seconds sub-unit scale in the driver, like other APIs pretend (e.g. counter api).

mbolivar-nordic commented:
Hi, I am circling back to older PRs now that the merge window is about to reopen. I want to help review the devicetree portions of this but I'm not familiar enough with RS485 to do a good job right now, so I'd like to ping @dcpleung as the UART maintainer to see if we can work together to get this in for v3.1.

dcpleung commented on drivers/serial/Kconfig:
```diff
@@ -59,6 +59,12 @@ config UART_USE_RUNTIME_CONFIGURE
 	  Say y if unsure. Disable this to reduce footprint for
 	  applications that do not require runtime UART configuration.
 
+config SERIAL_SUPPORT_RS485
+	bool
+	help
+	  This is an option to be enabled by individual serial driver
+	  to signal that the driver and hardware supports RS485.
+
```
I don't see this actually being used in code. So there is no need for this.

dcpleung reviewed (commented):
TBH, this is not exactly an API change as this does not introduce any new APIs (except the line control enum, but it is not a callable API).

The DTS looks good.

jfischer-no commented on dts/bindings/serial/uart-rs485-controller.yaml:
```diff
@@ -0,0 +1,41 @@
+    rs485-enabled:
+      type: boolean
+      required: false
+      description: Enable rs485 mode
```
Considering `UART_LINE_CTRL_RS485 = BIT(5),` and Kconfig option `SERIAL_SUPPORT_RS485`, this property does not seem necessary to me. Ideally, uart-rs485-controller.yaml should only describe bindings to a RS-485 transceiver. If Kconfig optoin `SERIAL_SUPPORT_RS485` is not enabled, attempt to set `UART_LINE_CTRL_RS485` should probably return an -ENOTSUP (API not enabled).

It would be nice to have optional bindings for the GPIOs if the controller does not have RS-485 and dedicated pins for DE and RE, see spi-controller.yaml as an example.

dleach02 commented:
@mbolivar-nordic can you revisit to determine if your change request has been addressed satisfactorily?
